plugins/workload-management/src/main/java/org/opensearch/plugin/wlm/rule/storage/DefaultAttributeValueStore.java

```diff
@@ -0,0 +1,103 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.plugin.wlm.rule.storage;
+
+import org.apache.commons.collections4.trie.PatriciaTrie;
+
+import java.util.Map;
+import java.util.Optional;
+
+/**
+ * This is a patricia trie based implementation of AttributeValueStore
+ * We are choosing patricia trie because it provides very fast search operations on prefix matches as well as range
+ * lookups. It provides a very efficient storage for strings
+ * ref: https://commons.apache.org/proper/commons-collections/javadocs/api-4.4/org/apache/commons/collections4/trie/PatriciaTrie.html
+ */
+public class DefaultAttributeValueStore<K extends String, V> implements AttributeValueStore<K, V> {
+    PatriciaTrie<V> trie;
+
+    /**
+     * Default constructor
+     */
+    public DefaultAttributeValueStore() {
+        this(new PatriciaTrie<>());
+    }
+
+    /**
+     * Main constructor
+     * @param trie A Patricia Trie
+     */
+    public DefaultAttributeValueStore(PatriciaTrie<V> trie) {
+        this.trie = trie;
+    }
+
+    @Override
+    public void add(K key, V value) {
+        trie.put(key, value);
+    }
+
+    @Override
+    public void remove(String key) {
+        trie.remove(key);
+    }
+
+    @Override
+    public Optional<V> get(String key) {
+        /**
+         * Since we are inserting prefixes into the trie and searching for larger strings
+         * It is important to find the largest matching prefix key in the trie efficiently
+         * Hence we can do binary search
+         */
+        String longestMatchingPrefix = findLongestMatchingPrefix(key);
+
+        /**
+         * Now there are following cases for this prefix
+         * 1. There is a Rule which has this prefix as one of the attribute values. In this case we should return the
+         *     Rule's label otherwise send empty
+         */
+        for (Map.Entry<String, V> possibleMatch : trie.prefixMap(longestMatchingPrefix).entrySet()) {
+            if (key.startsWith(possibleMatch.getKey())) {
+                return Optional.of(possibleMatch.getValue());
+            }
+        }
+
+        return Optional.empty();
+    }
+
+    private String findLongestMatchingPrefix(String key) {
+        int low = 0;
+        int high = key.length() - 1;
+
+        while (low < high) {
+            int mid = low + (high - low + 1) / 2;
```

Comment:
shouldn't this just be be (low + high) / 2 + 1? or (low+high+1)/2?

Reply:
Since the loop invariant is **(low < high)**  at some point `high = low + 1` in this case .  (low + high) can suffer from integer overflow and the implemented instruction avoids both of them.

`(low + high) / 2 + 1` is not right specially if the optimal value is low+1 when high = low + 2. In this case we will skip the `low + 1` point.

Reply:
> (low + high) can suffer from integer overflow and the implemented instruction avoids both of them.

(low + high) is bound by 2 * key.length(). If it can cause integer overflow, I will be worried about many other memory issues than this itself. I guess we can make good practical assumption that keys are not going to be bigger than `MAX_INT/2`, which is almost 1G already

Reply:
> (low + high) / 2 + 1 is not right specially if the optimal value is low+1 when high = low + 2. In this case we will skip the low + 1 point.

Yeah, I just quickly suggested one of the 2 options that should work instead of the expression we have currently. Did not verify the correctness for both

Reply:
I agree with this, but the expression was designed to be generic, We can replace that with (low + high + 1) / 2 since mathematically it is same do you think the new expression has any advantage over the one we have written apart from 1 additional arithmatic operation.

Reply:
I will make the suggested change.